types/index.d.ts

```diff
@@ -0,0 +1,166 @@
+// Minimum TypeScript Version: 4.2
+
+// NOTE: `dtslint` needs 4.2 even though these definitions should work with 4.1.
+
+declare module "hyperapp" {
+  // `app()` initiates a Hyperapp application. `app()` along with runners and
+  // subscribers are the only places where side effects are allowed.
+  function app<S>(props: App<S>): void
+
+  // `h()` builds a virtual DOM node.
+  function h<S, T extends string = string>(
+    // Tags cannot be empty strings.
+    tag: T extends "" ? never : T,
+    props: PropList<S>,
+    children?: VNode<S> | readonly VNode<S>[]
+  ): VDOM<S>
+
+  // `memo()` stores a view along with data for it.
+  function memo<
+    S,
+    D extends string | any[] | Record<string, any>
+  >(
+    view: View<D>,
+    data: D
+  ): VDOM<S>
+
+  // `text()` creates a virtual DOM node representing plain text.
+  function text<T, S>(
+    value: T extends (symbol | ((..._: any[]) => any)) ? never : T,
+    node?: Node
+  ): VDOM<S>
+
+  // ---------------------------------------------------------------------------
+
+  // A Hyperapp application instance has an initial state and a base view.
+  // It must also be mounted over an available DOM element.
+  type App<S> = Readonly<{
+    init: StateFormat<S> | Action<S>
+    view: View<S>
+    node: Node
+    subscriptions?: Subscriptions<S>
+    middleware?: Middleware<S>
+  }>
+
+  // A view builds a virtual DOM node representation of the application state.
+  type View<S> = (state: State<S>) => VDOM<S>
+
+  // The subscriptions function manages a set of subscriptions.
+  type Subscriptions<S> = (state: State<S>) => Subscription<S>[]
+
+  // A subscription represents subscriber activity.
+  type Subscription<S, D = any> = boolean | undefined | SubscriberDescriptor<S, D> | Unsubscribe
+
+  // A subscriber reacts to subscription updates.
+  type SubscriberDescriptor<S, D> = [Subscriber<S, D>, Payload<D>]
+  type Subscriber<S, D> = (dispatch: Dispatch<S>, props?: Payload<D>) => void | Unsubscribe
+
+  // An unsubscribe function cleans up a canceled subscription.
+  type Unsubscribe = () => void
+
+  // Middleware allows for custom processing during dispatching.
+  type Middleware<S> = (dispatch: Dispatch<S>) => Dispatch<S>
+
+  // ---------------------------------------------------------------------------
+
+  // A dispatched action handles an event in the context of the current state.
+  type Dispatch<S> = (action: Action<S>, props?: Payload<any>) => void
+
+  // An action transforms existing state and/or wraps another action.
+  type Action<S, P = any> = ActionTransform<S, P> | ActionDescriptor<S, P>
+  type ActionTransform<S, P = any> = (state: State<S>, props?: Payload<P>) => StateFormat<S> | Action<S>
+  type ActionDescriptor<S, P> = [ActionTransform<S, P>, Payload<P>]
+
+  // A transform carries out the transition from one state to another.
+  type Transform<S, P = any> = (state: StateFormat<S>, props?: Payload<P>) => StateFormat<S>
```

Comment:
What's a `Transform`? 💭

Reply:
`Transform` would apply to the functions that more complicated actions would leverage.

Reply:
What do you mean by a  small action? Actions can't be called, only given to Hyperapp or returned from other actions.

Reply:
I don't see much value to defining this type tbh. I mean sure, it exists as a concept but it could easily be defined by a user who needs it.

Reply:
@jorgebucaran In the docs [here](https://github.com/jorgebucaran/hyperapp/blob/main/docs/architecture/actions.md#transforms), the `Liokaiser` action calls functions that would, in a TypeScript project, be defined in terms of `Transform`.  A small action would be small enough to not have to be split up.

@zaceno There are was a Slack discussion about `Transform` that you had at the time agreed with me on:  https://hyperapp.slack.com/archives/C41ECC0V6/p1600930431006000?thread_ts=1600929373.003100&cid=C41ECC0V6

My defense for having `Transform` ready "out-of-the-box" is that I had found it to be a useful type that would show up often in larger projects.